libr/anal/type.c

```diff
@@ -330,35 +279,81 @@ static RAnalBaseType *get_union_type(RAnal *anal, const char *sname) {
 	return NULL;
 }
 
+static RAnalBaseType *get_typedef_type(RAnal *anal, const char *sname) {
+	r_return_val_if_fail (anal && R_STR_ISNOTEMPTY (sname), NULL);
+
+	RAnalBaseType *base_type = r_anal_base_type_new (R_ANAL_BASE_TYPE_KIND_TYPEDEF);
+	if (!base_type) {
+		return NULL;
+	}
+
+	base_type->type = get_type_data (anal->sdb_types, "typedef", sname);
+	if (!base_type->type) {
+		goto error;
+	}
+	return base_type;
+
+error:
+	r_anal_base_type_free (base_type);
+	return NULL;
+}
+
+static RAnalBaseType *get_atomic_type(RAnal *anal, const char *sname) {
+	r_return_val_if_fail (anal && R_STR_ISNOTEMPTY (sname), NULL);
+
+	RAnalBaseType *base_type = r_anal_base_type_new (R_ANAL_BASE_TYPE_KIND_ATOMIC);
+	if (!base_type) {
+		return NULL;
+	}
+
+	base_type->type = get_type_data (anal->sdb_types, "type", sname);
+	if (!base_type->type) {
+		goto error;
+	}
+
+	RStrBuf key;
+	r_strbuf_init (&key);
+	r_strbuf_setf (&key, "type.%s.size", sname);
+	base_type->size = sdb_num_get (anal->sdb_types, r_strbuf_get (&key), 0);
+	r_strbuf_fini (&key);
+
+	return base_type;
+
+error:
+	r_anal_base_type_free (base_type);
+	return NULL;
+}
+
 // returns NULL if name is not found or any failure happened
 R_API RAnalBaseType *r_anal_get_base_type(RAnal *anal, const char *name) {
 	r_return_val_if_fail (anal && name, NULL);
 
 	char *sname = r_str_sanitize_sdb_key (name);
 	const char *type = sdb_const_get (anal->sdb_types, sname, NULL);
-
-	// Right now just types: struct, enum, union are supported
-	if (!type || !(strcmp (type, "enum") || strcmp (type, "struct") || strcmp (type, "union"))) {
+	if (!type) {
 		free (sname);
 		return NULL;
 	}
-	// Taking advantage that all 3 types start with distinct letter
-	// because the strcmp condition guarantees that only those will get to this flow
-	RAnalBaseType *base_type = NULL;
 
-	switch (type[0]) {
-	case 's': // struct
+	RAnalBaseType *base_type = NULL;
+	if (!strcmp (type, "struct")) {
 		base_type = get_struct_type (anal, sname);
-		break;
-	case 'e': // enum
+	} else if (!strcmp (type, "enum")) {
 		base_type = get_enum_type (anal, sname);
-		break;
-	case 'u': // union
+	} else if (!strcmp (type, "union")) {
 		base_type = get_union_type (anal, sname);
-		break;
+	} else if (!strcmp (type, "typedef")) {
+		base_type = get_typedef_type (anal, sname);
+	} else if (!strcmp (type, "type")) {
```

Comment:
uhm how that would look in C? i mean for the cparse side.

Reply:
cparse as in parsing a type string like "char **[]" or parsing a C header? Atomic types can't be created in C, they are only for specifying the exact properties of `char`, `int`, etc. and usually you will have them loaded by default already.

Reply:
for me atomic type have a different meaning, which is basically data types that cant introduce race codnitions. not sure if you are refering to the same or you are just talking about the basic types which have nothing to do with atomic. in that case i would suggest a different name.

Reply:
I think I was the one who named them atomic types, as basic types would kinda collide, I thought of atomic as a further undivisable (internally unstructured).

Reply:
I think atomic is actually a very good name here. "atomic" indeed can refer to the multi-threading things, but imo having this meaning for a typekind here doesn't make much sense so it kind of rules itself out.
Instead, "atomic" or "atoms" is also a very common term for entities in formulas that can't be split up further: https://en.wikipedia.org/wiki/Atomic_formula
which is pretty much exactly what we have here.